### DISCUSSION TITLE
Does lua.LState safe to reuse for running different code?

### DISCUSSION THREAD

**Discussion by wyxloading** [yuin/gopher-lua#391] [Q&A]
- [x] GopherLua is a Lua5.1 implementation. You should be familiar with Lua programming language. Have you read [Lua 5.1 reference manual](http://www.lua.org/manual/5.1/) carefully?
- [x] GopherLua is a Lua5.1 implementation.  In Lua, to keep it simple, **it is more important to remove functionalities rather than to add functionalities unlike other languages** . If you are going to introduce some new cool functionalities into the GopherLua code base and the functionalities can be implemented by existing APIs, It should be implemented as a library. 

Please answer the following before submitting your issue:

1. What version of GopherLua are you using? : master
2. What version of Go are you using? : go1.14.1
3. What operating system and processor architecture are you using? : linux amd64
4. What did you do? : 
```golang
package main

import (
	"errors"

	"github.com/yuin/gluare"
	lua "github.com/yuin/gopher-lua"
	luar "layeh.com/gopher-luar"
)

func main() {
	l := lua.NewState(lua.Options{
		SkipOpenLibs: true,
	})
	for _, pair := range []struct {
		n string
		f lua.LGFunction
	}{
		{lua.LoadLibName, lua.OpenPackage}, // Must be first
		{lua.BaseLibName, lua.OpenBase},
		{lua.TabLibName, lua.OpenTable},
		{lua.StringLibName, lua.OpenString},
		{lua.MathLibName, lua.OpenMath},
	} {
		if err := l.CallByParam(lua.P{
			Fn:      l.NewFunction(pair.f),
			NRet:    0,
			Protect: true,
		}, lua.LString(pair.n)); err != nil {
			l.Close()
			panic(err)
		}
	}
	l.PreloadModule("re", gluare.Loader)


	for i := 0; i < 10; i++ {
		_, err := reuseLState(l)
		if err != nil {
			panic(err)
		}
	}
}

var (
	code =
`function mytest(params)
	return "a"
end`

	errReturnNotExpectd = errors.New("return not expected")
)

func reuseLState(ls *lua.LState) (string, error) {
	err := ls.DoString(code)
	if err != nil {
		ls.Close()
		return "", err
	}
	fn := ls.GetGlobal("mytest")
	if fn.Type() != lua.LTFunction {
		ls.Close()
		return "", err
	}
	lp := lua.P{
		Fn:      fn,
		NRet:    1,
		Protect: true,
	}
	params := map[string]interface{}{
		"key1": "value",
		"key2": 123,
	}
	if err := ls.CallByParam(lp, luar.New(ls, params)); err != nil {
		return "", err
	}
	ret := ls.Get(-1)
	ls.Pop(1) // not sure how many i should pop
	if ls, ok := ret.(lua.LString); !ok {
		return "", errReturnNotExpectd
	} else {
		return ls.String(), nil
	}
}

```

5. What did you expect to see? :
reuse the `lua.LState` and keep the memory usage low ( without the code, params )

Could anyone provide an example about how to reuse lua.LState safely?

**Comment by kulti** [comment]
Hi. I reuse my lua states by the following:
```
	oldTop := w.luaState.GetTop()
	defer w.luaState.SetTop(oldTop)
```

I'm not 100% sure that is right way, but it helps me.
